### PR TITLE
WIP: Adding GDB stub server for EE/IOP 

### PR DESCRIPTION
**Pull request by vertver**
The basic functionality of the GDB server is currently implemented, but it requires extended support from the GDB client (building binutils with libexpat support). This code is currently WIP, so I don't think it can be merged anytime soon.

### CODE REVIEW

**Comment by Ziemas** [comment]
Thank you for taking a stab at this, it'd be extremely awesome to have!

I couldn't resist playing around a bit so here's a patch that fixes building and connection linux connect (albeit eventually faling with `Received too much data from the target.` for some reason.

```diff
From 4edffcd0c4bc894eb07b48c58249a06dc88a3a47 Mon Sep 17 00:00:00 2001
From: Ziemas <ziemas@ziemas.se>
Date: Wed, 24 May 2023 07:15:48 +0200
Subject: [PATCH] DebugServer: Fix linux

---
 pcsx2/DebugTools/DebugServer.cpp | 42 +++++++++++++++++---------------
 1 file changed, 23 insertions(+), 19 deletions(-)

diff --git a/pcsx2/DebugTools/DebugServer.cpp b/pcsx2/DebugTools/DebugServer.cpp
index 526d64b815..1f06c33de6 100644
--- a/pcsx2/DebugTools/DebugServer.cpp
+++ b/pcsx2/DebugTools/DebugServer.cpp
@@ -35,6 +35,9 @@ DebugNetworkServer VU1DebugNetworkServer;
 #include <sys/socket.h>
 #include <sys/un.h>
 #include <unistd.h>
+#include <fcntl.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
 #endif
 
 DebugNetworkServer::DebugNetworkServer()
@@ -84,8 +87,6 @@ DebugNetworkServer::shutdown()
 	{
 #ifdef _WIN32
 		WSACleanup();
-#else
-		unlink(m_socket_name.c_str());
 #endif
 
 		close_portable(m_sock);
@@ -134,7 +135,7 @@ DebugNetworkServer::reviveConnection()
 {
 	if (m_msgsock != 0)
 	{
-		closesocket(m_msgsock);
+		close_portable(m_msgsock);
 		m_msgsock = 0;
 	}
 
@@ -149,15 +150,14 @@ DebugNetworkServer::reviveConnection()
 		m_msgsock = accept(m_sock, 0, 0);
 #ifdef _WIN32
 		int errno_w = WSAGetLastError();
-		bool notFailed = (errno_w == WSAECONNRESET || errno_w == WSAEINTR || errno_w == WSAEINPROGRESS || errno_w == WSAEMFILE || errno_w == WSAEWOULDBLOCK);
-#else
-		bool notFailed = (errno == ECONNABORTED || errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK);
-#endif
 		if (errno_w == 0)
 		{
 			break;
 		}
-
+		bool notFailed = (errno_w == WSAECONNRESET || errno_w == WSAEINTR || errno_w == WSAEINPROGRESS || errno_w == WSAEMFILE || errno_w == WSAEWOULDBLOCK);
+#else
+		bool notFailed = (errno == ECONNABORTED || errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK);
+#endif
 		if (notFailed)
 		{
 			Threading::Sleep(1);
@@ -170,14 +170,13 @@ DebugNetworkServer::reviveConnection()
 
 	} while (m_msgsock == (u64)-1);
 
-	u_long mode = 1;
 #ifdef _WIN32
+	u_long mode = 1;
 	if (ioctlsocket(m_msgsock, FIONBIO, &mode) < 0)
 #else
-	if (fcntl(m_msgsock, O_NONBLOCK, &mode) < 0)
+	if (fcntl(m_msgsock, F_SETFL, fcntl(m_msgsock, F_GETFL) | O_NONBLOCK) < 0)
 #endif
 	{
-		int errno_w = WSAGetLastError();
 		Console.WriteLn(Color_Red, "DebugNetworkServer: unnable to set socket as non-blocking! Shutting down...");
 		return false;
 	}
@@ -217,31 +216,36 @@ DebugNetworkServer::serverLoop()
 		return;
 	}
 #else
-	struct sockaddr_un server;
-	m_sock = socket(AF_UNIX, SOCK_STREAM, 0);
+	struct sockaddr_in server;
+	m_sock = socket(AF_INET, SOCK_STREAM, 0);
 	if (m_sock < 0)
 	{
 		Console.WriteLn(Color_Red, "DebugNetworkServer: Cannot open socket! Shutting down...");
 		shutdown();
-		return false;
+		return;
 	}
 
-	server.sun_family = AF_UNIX;
+	server.sin_family = AF_INET;
+	server.sin_addr.s_addr = 0;
+	if (!m_address.empty()) {
+		inet_pton(AF_INET, m_address.data(), &(server.sin_addr));
+	}
+	server.sin_port = htons(m_port);
+
 	if (bind(m_sock, (struct sockaddr*)&server, sizeof(struct sockaddr_un)))
 	{
 		Console.WriteLn(Color_Red, "DebugNetworkServer: Error while binding to socket! Shutting down...");
 		shutdown();
-		return false;
+		return;
 	}
 #endif
-	u_long mode = 1;
 #ifdef _WIN32
+	u_long mode = 1;
 	if (ioctlsocket(m_sock, FIONBIO, &mode) < 0)
 #else
-	if (fcntl(m_sock, O_NONBLOCK, &mode) < 0)
+	if (fcntl(m_msgsock, F_SETFL, fcntl(m_msgsock, F_GETFL) | O_NONBLOCK) < 0)
 #endif
 	{
-		int errno_w = WSAGetLastError();
 		Console.WriteLn(Color_Red, "DebugNetworkServer: unnable to set socket as non-blocking! Shutting down...");
 	}
 
-- 
2.40.1

```

**Comment by vertver** [comment]
@Ziemas, yeah it might be a problem with a packet size. Since we're sending a whole XML file containing full information about current CPU, it might be helpful to change line 252 in GDBServer.cpp from this...
```cpp
	for (std::size_t cat = 0; cat < cpuInterface->getRegisterCategoryCount(); cat++)
```
to this:
```cpp
	for (std::size_t cat = 0; cat < 1/*cpuInterface->getRegisterCategoryCount()*/; cat++)
```

This is a temporary fix, so there must be a more pleasant solution for this. Anyway, even if you will apply this fix - you need to rebuild binutils with libexpat (or xml) support (just because it's only way to define custom platforms in GDB, check this: https://sourceware.org/gdb/onlinedocs/gdb/Target-Description-Format.html#Target-Description-Format).

**Comment by Ziemas** [comment]
Looking at pcsx-redux which has a working gdb server implementation I see that it has this WritePaged implementation that it uses for sending the XML.

https://github.com/grumpycoders/pcsx-redux/blob/main/src/core/gdb-server.cc#L394-L403

Maybe something like that is needed.

File: pcsx2-qt/Settings/DebugSettingsWidget.ui [hunk 1]
```diff
@@ -34,6 +34,118 @@
...
+                     <sizepolicy hsizetype="Minimum" vsizetype="Fixed">
+                      <horstretch>0</horstretch>
+                      <verstretch>0</verstretch>
+                     </sizepolicy>
+                    </property>
+                  </widget>
+			    </item>
+		      </layout>
+		   </widget>
+	   </item>
```
Comment (lightningterror) [inline pcsx2-qt/Settings/DebugSettingsWidget.ui:146]: Formatting, above changes as well.

**Comment by refractionpcsx2** [comment]
Adding Hold Merge so nobody accidentally merges this before it's ready

**Comment by refractionpcsx2** [comment]
Needs a rebase as the patches have changed so any new builds will fail unless a rebase is done.

**Comment by vertver** [comment]
PR was rebased on ed6dd6f6cdb7a41c35ea18e88625a45badb6f9e5 commit

**Comment by refractionpcsx2** [comment]
Awesome, thanks :)

**Comment by ivandortulov** [comment]
@vertver Can you please rebase this again? The build has expired and I cannot trigger a new one. Building it locally is too complicated for me as I am using a Windows environment and I would like to avoid it.

**Comment by vertver** [comment]
@ivandortulov, this implementation is not complete, it's just implementing basic [GDB stub](https://sourceware.org/gdb/current/onlinedocs/gdb.html/Packets.html#bc) without required features. (features as step-out, continue, thread info and others are still don't exist). If you want to continue work on this PR - sure, I'll rebase it.

**Comment by refractionpcsx2** [comment]
So you've given up on this PR? if this is just going to go stale, that's a shame, but there's not much point in leaving it open.

**Comment by vertver** [comment]
@refractionpcsx2, was going to revisit this PR but have been very busy. I think I'm free now and might be able to do something about it. I'll rebase and continue the implementation in a few days.

**Comment by refractionpcsx2** [comment]
Okay, thank you.

**Comment by Ziemas** [comment]
You can do an interactive rebase and drop all the the commits except yours.

**Comment by vertver** [comment]
Rebased, I will Open new PR as soon as it is ready for use.

**Comment by ivandortulov** [comment]
Thank you

**Pull request closed without merging**